### PR TITLE
Update tests for Python 3

### LINKED ISSUE #82: Update tests for Python 3
tests for memcached written in Python 2, but Python 2 is deprecated.
test-run supports Python 3 [1] and it would be better to update tests for Python 3.

1: https://github.com/tarantool/test-run/issues/20

### PR DESCRIPTION
**Pull request by ligurio**
Closes #82

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,10 +14,11 @@ Memcached protocol 'wrapper' for Tarantool.
 
  * Tarantol 1.6.8+ with header files (tarantool && tarantool-dev packages).
  * Cyrus SASL library (with header files)
- * Python >= 2.7, <3 with next packages (for testing only):
+ * Python >= 2.7 with next packages (for testing only):
```
Comment (LeonidVas) [inline README.md:17]: Maybe Python >= 3 is better because we are testing with Python3?

**Review by LeonidVas — APPROVED** [review]

File: test/binary/suite.ini [hunk 2]
```diff
@@ -2,4 +2,4 @@
 core = tarantool
 script = binary.lua
 description = memcached binary tests
-disabled = binary-boundary.test.py
+disabled = binary-expire.test.py binary-boundary.test.py
```
Comment (Totktonada) [inline test/binary/suite.ini]: I guess you had an idea to disable it and return here later, but then you found a fix: the next commit enables the test back. So, it seems, the commit that disables the test is not needed anymore?

File: README.md [hunk 3]
```diff
@@ -14,10 +14,11 @@ Memcached protocol 'wrapper' for Tarantool.
...
    - PyYAML
    - msgpack-python
    - six==1.9.0
+   - gevent==1.1b5
```
Comment (Totktonada) [inline README.md:21]: This list is not clear for me. At least it is not based on actual test-run's `requirements.txt`: we don't depend on external `msgpack-python` (it is bundled into test-run), we don't depend on `six`, we have `gevent==21.*` (but any gevent version fit good for test-run in fact if it works on given Python version).

File: .github/workflows/testing.yaml [hunk 4]
```diff
@@ -33,10 +33,10 @@ jobs:
```
Comment (Totktonada) [inline .github/workflows/testing.yaml:33]: The `test/var/memcached` symlink should be removed together with test-run update, I think, because otherwise it is deleted by a first in-source run of tests: it is considered by `git status` as a tracked file change.

File: test/internal/compat.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import sys
+
+# Added for compatibility with Python 3.
+# Python 2 provided the class basestring, from which both str and unicode
+# derived. In Python 3, the concept of basestring makes no sense: text is only
+# represented by str.
+try:
+  basestring
+except NameError:
+  basestring = str
```
Comment (Totktonada) [inline test/internal/compat.py:10]: We don't use `basestring` directly anywhere, only via `string_types`. So we don't use `basestring` when running on Python 3. I removed this chunk of the code and all enabled tests succeeded.

File: test/binary/binary-expire.test.py [hunk 6]
```diff
@@ -41,15 +41,15 @@ def check(key, flags, val, level = 0):
 
 stat = mc.stat("reset")
 
-for i in xrange(10000):
+for i in range(10000):
     mc.set('key-%d' % i, 'value-%d' % i, expire=1)
 
 stat = mc.stat()
-while int(stat.get('evictions', '0')) < 10000:
+while int(stat.get(b'evictions', '0')) < 10000:
```
Comment (Totktonada) [inline test/binary/binary-expire.test.py:48]: It seems convenient to convert keys into strings (`str` type disregarding Python version) right in `mc.stat()`. At least python-binary-memcached does it: see `protocol.py`.

Feel free to resolve without changes, this is cosmetic.

File: test/internal/memcached_connection.py [hunk 7]
```diff
@@ -238,7 +239,10 @@ def _read_and_parse_response(self):
             if is_indecrq(op):
                 val = INDECR_STRUCT.unpack_from(val)[0]
         if val is not None:
-            retval['val'] = val
+            if isinstance(val, bytes):
+                retval['val'] = bytes_to_str(val)
+            else:
+                retval['val'] = val
```
Comment (Totktonada) [inline test/internal/memcached_connection.py:245]: Feel free to resolve this comment without changes too, but I'll tell what I think here.

[python-memcached](https://github.com/linsomniac/python-memcached) and [python-binary-memcached](https://github.com/jaysonsantos/python-binary-memcached) set a test/binary flag on writting a value and lean on the flag to decode into bytes or text at reading. It is quite intuitive for Python's client. However it is not wortful to implement for our testing implementation.

So here we agree on unconditional interpreting values as text ones. It would be good to mention this fact somewhere.

Next, why I even looked at this hunk of the code? Here we don't use our understanding of existing code to make the new code simpler. What I can deduce from your variant of the code? `val` is either bytes or something else. But what else? A string? Hm... I need to read `_read_response()` and `_recv()` to understand what is going on here.

If we'll rewrite it in a more strict way:

```python
if val_lenv is None or val_len > 0:
    begin = HEADER_SIZE + ext_len + key_len
    end   = HEADER_SIZE + tot_len
    val   = cmd[begin:end] 
    if is_indecrq(op): 
        # incr[q]/decr[q] result is integer.
        val = INDECR_STRUCT.unpack_from(val)[0]
    else:
        # Ideally we should lean on 4th bit in flags to
        # decide whether to interpret this value as binary
        # or as text.
        # 
        # Unconditional interpreting it as a text is
        # enough for testing purposes, though.
        val = bytes_to_str(val)
    retval['val'] = val
```

Everything becomes simple. If there is a value in the answer, we decode it as integer for increment / decrement and decode it as a text otherwise. No need to bury into other code.

It is stylistic, of course, but interesting. Ever if you'll not change it, please, share, what do you think?

File: test/binary/binary-boundary.test.py [hunk 8]
```diff
@@ -10,14 +10,16 @@
...
 mc = MemcachedBinaryConnection("127.0.0.1", iproto.py_con.port)
 
 def iequal(left, right, level = 1):
     if (left != right):
         tb = traceback.extract_stack()[-(level + 1)]
-        print "Error on line %s:%d: %s not equal %s" % (tb[0], tb[1],
-                repr(left), repr(right))
-        if (isinstance(left, basestring)):
+        print("Error on line %s:%d: %s not equal %s" % (tb[0], tb[1],
+                repr(left), repr(right)))
```
Comment (Totktonada) [inline test/binary/binary-boundary.test.py:21]: To be honest, it would be a bit easier to look over the changes if all those print -> print() ones would be in a separate commit. I already looked over, so it is not important for me to review. Feel free to resolve the comment without changes and save some time.

**Review by Totktonada — APPROVED** [review]
I see nothing criminal :) Feel free to push.

I left some stylistic thoughts. I don't know, where you'll find youself in a best balance between effort and result, but I suggest to stop at this point and go ahead :)

File: README.md (hunk 1 above)
Comment (ligurio) [reply README.md:17]: To be honestly, we need to specify here all supported Python versions. However, 2.x is EOL, and we can limit versions by 3.x. So I replaced to ">= 3.7" (Python 3.7 is a minimal supported version, see https://endoflife.date/python):

```diff
--- a/README.md
+++ b/README.md
@@ -14,7 +14,7 @@ Memcached protocol 'wrapper' for Tarantool.
 
  * Tarantol 1.6.8+ with header files (tarantool && tarantool-dev packages).
  * Cyrus SASL library (with header files)
- * Python >= 2.7 with next packages (for testing only):
+ * Python >= 3.7 with next packages (for testing only):
    - PyYAML
    - msgpack-python
    - six==1.9.0
```

File: test/binary/suite.ini (hunk 2 above)
Comment (ligurio) [reply test/binary/suite.ini]: Right, forgot to remove it. Thanks!

File: README.md (hunk 3 above)
Comment (ligurio) [reply README.md:21]: Fixed:

```diff
--- a/README.md
+++ b/README.md
@@ -16,10 +16,8 @@ Memcached protocol 'wrapper' for Tarantool.
  * Tarantol 1.6.8+ with header files (tarantool && tarantool-dev packages).
  * Cyrus SASL library (with header files)
  * Python >= 3.7 with next packages (for testing only):
-   - PyYAML
-   - msgpack-python
-   - six==1.9.0
-   - gevent==1.1b5
+   - PyYAML 5+
+   - gevent 21+
 
 ### Installation
 
```

File: test/binary/binary-boundary.test.py (hunk 8 above)
Comment (ligurio) [reply test/binary/binary-boundary.test.py:21]: done

File: .github/workflows/testing.yaml (hunk 4 above)
Comment (ligurio) [reply .github/workflows/testing.yaml:33]: done

File: test/internal/compat.py (hunk 5 above)
Comment (ligurio) [reply test/internal/compat.py:10]: Right, removed.

File: test/binary/binary-expire.test.py (hunk 6 above)
Comment (ligurio) [reply test/binary/binary-expire.test.py:48]: Updated, thanks!

```diff
diff --git a/test/binary/binary-expire.test.py b/test/binary/binary-expire.test.py
index f8f293f..50c2698 100644
--- a/test/binary/binary-expire.test.py
+++ b/test/binary/binary-expire.test.py
@@ -45,11 +45,11 @@ for i in range(10000):
     mc.set('key-%d' % i, 'value-%d' % i, expire=1)
 
 stat = mc.stat()
-while int(stat.get(b'evictions', '0')) < 10000:
+while int(stat.get('evictions', '0')) < 10000:
     time.sleep(0.01)
     stat = mc.stat()
 
-issert(b'evictions' in stat)
-iequal(int(mc.stat().get(b'evictions', 0)), 10000)
+issert('evictions' in stat)
+iequal(int(mc.stat().get('evictions', 0)), 10000)
 
 sys.path = saved_path
diff --git a/test/binary/binary.test.py b/test/binary/binary.test.py
index 319eb1b..51a77fc 100644
--- a/test/binary/binary.test.py
+++ b/test/binary/binary.test.py
@@ -251,7 +251,7 @@ empty("x")
 empty("y")
 
 stat2 = mc.stat()
-iequal(int(stat1[b'cmd_flush']) + 1, int(stat2[b'cmd_flush']))
+iequal(int(stat1['cmd_flush']) + 1, int(stat2['cmd_flush']))
 
 print("""#----------------------------# diagnostics append #---------------------------#""")
 key, value, suffix = "appendkey", "some value", " more"
diff --git a/test/internal/memcached_connection.py b/test/internal/memcached_connection.py
index 2b742fa..a707f1b 100644
--- a/test/internal/memcached_connection.py
+++ b/test/internal/memcached_connection.py
@@ -481,7 +485,7 @@ class MemcachedBinaryConnection(TarantoolConnection):
             if 'key' in rv and not rv['key'] and \
                'val' in rv and not rv['val']:
                 return ans
-            ans[rv['key']] = rv['val']
+            ans[bytes_to_str(rv['key'])] = rv['val']
         return ans
 
     @binary_decorator
```

File: test/internal/memcached_connection.py (hunk 7 above)
Comment (ligurio) [reply test/internal/memcached_connection.py:245]: Updated, thanks!

```diff
diff --git a/test/internal/memcached_connection.py b/test/internal/memcached_connection.py
index 2b742fa..a707f1b 100644
--- a/test/internal/memcached_connection.py
+++ b/test/internal/memcached_connection.py
@@ -238,11 +238,15 @@ class MemcachedBinaryConnection(TarantoolConnection):
             # decode result of (incr/decr)(q)
             if is_indecrq(op):
                 val = INDECR_STRUCT.unpack_from(val)[0]
-        if val is not None:
-            if isinstance(val, bytes):
-                retval['val'] = bytes_to_str(val)
             else:
-                retval['val'] = val
+                # Ideally we should lean on 4th bit in flags to
+                # decide whether to interpret this value as binary
+                # or as text.
+                #
+                # Unconditional interpreting it as a text is
+                # enough for testing purposes, though.
+                val = bytes_to_str(val)
+            retval['val'] = val
 
         return retval
 
```

File: test/binary/binary-boundary.test.py [hunk 9]
```diff
@@ -10,13 +10,15 @@
 from internal.memcached_connection import MemcachedBinaryConnection
 from internal.memcached_connection import STATUS, COMMANDS
 
+from internal.compat import string_types
+
```
Comment (Totktonada) [inline test/binary/binary-boundary.test.py:14]: Sorry for nitpicking, but it looks irrelevant to the 'test: convert print to function' commit.

Comment (ligurio) [reply test/binary/binary-boundary.test.py:14]: Right, fixed. Thanks!

**Comment by ligurio** [comment]
```
[001] binary/binary.test.py                                           [ fail ]
[001] 
[001] Test failed! Result content mismatch:
[001] --- binary/binary.result	Mon Jan 17 08:14:55 2022
[001] +++ var/rejects/binary/binary.reject	Mon Jan 17 08:15:35 2022
[001] @@ -15,8 +15,6 @@
[001]  #------------------------------# test decrement #-----------------------------#
[001]  #---------------------------------# bug 220 #---------------------------------#
[001]  #----------------------------------# bug 21 #---------------------------------#
[001] -#----------------------------------# bug 48 #---------------------------------#
[001] -Error on line binary/binary.test.py:168: 0 not equal 1
[001]  #-----------------------------------------------------------------------------#
[001]  ################################ Diagnostics CAS ##############################
[001]  #-----------------------------------------------------------------------------#
[001] 
```

**Review by Totktonada — APPROVED** [review]
